runtests: refactor and add lockfile option

hzhou opened the pull request:
## Pull Request Description

This PR adds a `lock=name` option to tests so big tests can be configured to avoid running simultaneously.

The larger commit is a refactor so adding new options to runtests are easier.
<!--
Insert description of the work in this merge request (above this comment),
particularly focused on _why_ the work is necessary, not _what_ you did.
-->

<!-- AUTHOR: After creating this merge request, check off each of the following items as you complete them. -->

## Expected Performance Changes

## Known Issues

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Passes tests (included warning check)
* [x] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [x] Commits are self-contained and do not do two things at once
* [x] Remove xfail from the test suite when fixing a test
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

hzhou commented:
test:jenkins/ch4/ofi

hzhou commented:
test:mpich/custom
testlist: attr
testlist: errhan
testlist: spawn

hzhou commented:
test:mpich/custom
compiler=intel
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:mpich/custom
compiler=intel
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:mpich/custom
compiler=intel
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:mpich/custom
compiler=intel
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
@raffenet Here is a successful run on a single Jenkins node with 10 concurrent workers:
https://pmrs-jenkins.cels.anl.gov/view/private/job/duplicate-review-ch4/15/

raffenet commented on test/mpi/maint/dtp-test-config.txt:
```diff
@@ -13,45 +13,40 @@
 attr/fkeyvaltype::1::1:1024:
-coll/bcast::1 512 262144:timeLimit=420:4:16:1024
-coll/bcast_comm_world_only::1 512 262144:timeLimit=360:10:16:1024
+coll/bcast::1 512::4:16:1024
```
Should the comment above be amended to describe the `mem=` argument?

hzhou replied:
Sure.

raffenet commented on test/mpi/maint/jenkins/xfail.conf:
```diff
@@ -38,10 +38,6 @@
 
-################################################################################
-# Add lock option to tests that require large memory
-* * * * * sed -i "s+\(^large_type_sendrec .*\)+\1 lock=mem+g" test/mpi/datatype/testlist
```
I presume we should just remove the commit that was added and then reverted before merging?

hzhou replied:
Of course. I'll clean it up.

raffenet commented on test/mpi/runtests.in:
```diff
@@ -1000,15 +1000,17 @@ sub CleanUpAfterRun {
-        print STDERR "Some programs ($programname) may still be running:\npids = ";
-        for (my $i=0; $i <= $#stillRunning; $i++ ) {
-            print STDERR $stillRunning[$i] . " ";
+        if ($verbose) {
```
I think this was already merged in another PR. I'm surprised Github isn't showing a conflict, but maybe it can be removed in the next update?

hzhou replied:
Yes. The other PR was cherry-picked from here. Git will silently drop duplicate/noop patches, therefore it does not get treated as conflict. A rebase will remove this commit.

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
All updated ready to be reviewed again.

raffenet commented on test/mpi/maint/dtp-test-config.txt:
```diff
@@ -13,45 +13,40 @@
+coll/bcast::1 512::4:16:1024
+coll/bcast::262144:timeLimit=420 mem=1.2:4:16:
+coll/bcast_comm_world_only::1 512::10:16:1024
+coll/bcast_comm_world_only::262144:timeLimit=360 mem=1.9:10:16:
```
I thought we discussed modifying the comment header in this file to describe the `mem=` format?

hzhou replied:
I remember I did. But apparently I was somehow distracted and the commit (amend) seems to be lost. Anyway, added a new commit for that.

hzhou commented:
test:jenkins/ch4/ofi

hzhou commented:
@raffenet  Could you give this another go?

raffenet commented on test/mpi/runtests.in:
```diff
@@ -769,9 +780,32 @@ sub RunMPIProgram {
+
+    # acquire lock if requested
+    my $lockfile="/tmp/runtests-mem.lock";
+    my ($lock_sh, $lock_ex) = (1, 2);
```
how is the regular lock intended to be used? i only see memory locking in this PR. is that right?

hzhou replied:
It is enabled directly by setting `lock=name` in the testlines. Comment added.

raffenet commented on test/mpi/runtests.in:
```diff
@@ -769,9 +780,32 @@ sub RunMPIProgram {
+        if ($test_opt->{mem} > $g_opt->{memory_total}) {
+            SkippedTest($programname, $np, $test_opt, $curdir, "xfail due to memory requirement");
+            next;
+        } elsif ($test_opt->{mem} * $g_opt->{memory_multiplier} > $g_opt->{memory_total} ) {
```
I see memory multiplier initialized to one. How/when is it updated?

hzhou replied:
It is set by environment variable `MPITEST_MEMORY_TOTAL` and `MPITEST_MEMORY_MULTIPLIER`, ref line 150.

raffenet commented on test/mpi/maint/jenkins/xfail.conf:
```diff
@@ -30,19 +30,6 @@
-################################################################################
-# other failures on FreeBSD because of memory limits in our jenkins
-# infrastructure
-* * * * freebsd64 sed -i "s+\(^large_type_sendrec .*\)+\1 xfail=ticket0+g" test/mpi/datatype/testlist
```
we'll have to make sure to test this with `--enable-large-tests`, which is currently set to "no" for review jobs.

hzhou replied:
Hmm... it seems the it is enabled even for review tests.

hzhou commented on test/mpi/runtests.in:
```diff
@@ -1332,3 +1372,30 @@ sub SkippedTest {
+        print "Taking lock [$lockfile], type $lock_type\n";
+    }
+    if(open LOCK, ">$lockfile"){
+        flock(LOCK, $lock_type);
```
Comments added.

hzhou commented on test/mpi/runtests.in:
```diff
@@ -769,9 +780,32 @@ sub RunMPIProgram {
     print STDOUT "Env includes $progEnv\n" if $verbose;
     print STDOUT "$mpiexec $np_arg $np $extraArgs $mpiexecArgs $program_wrapper ./$programname $progArgs\n" if $verbose;
     print STDOUT "." if $showProgress;
+
```
Cleared.

hzhou commented:
@raffenet I have addressed your comments and also made a few misc changes. I pushed an additional commit so you don't need to review from beginning again. Take another look?
